### ISSUE TITLE
Provide a way to disable non hashed CSS classes

### ISSUE THREAD

**Issue by vandop** [OfficeDev/office-ui-fabric-react#1329]
When having building your own component that is integrated in webpages that also use Fabric UI we are getting into collision madness.

Our temporary solution is to increase CSS specificity all over the place, but that makes our solution very brittle and whoever owns the webpage can easily break us by introducing one more property that we are not overriding.

We understand that there is a plan long term, but we  believe that a short/midterm solution could be applied, by giving the chance to not stamp non-hashed classes to the components through a property.

For instance:

```jsx
<CommandButton
          data-automation-id='test'
          icon='ContactCard'
          disabled={ disabled }
          useOnlyHashedClassNames={true}>
          Create account
</CommandButton>
```

I can create a pull request for this case and help out with this.

**Comment by dzearing** [comment]
To understand the issue better;

1. You render a `CommandButton` which injects `ms-Button` classname in the root.
2. The app overrides .ms-Button, affecting your ux inadvertently causing bugs.

To resolve, you'd like a feature on CommandButton to render without ms-Button:

`withoutGlobalClasses: boolean`

Of course passing this in recursively everywhere is going to be a mess. For this to work well I think we need to fast track some of the theming work that we're investigating.

Basically we want a way to override component defaults without modifying the app code:

```tsx
<ThemeProvider theme={{
  componentDefaults: {
    CommandButton: {
      withoutGlobalClasses: true
    }
  }
}}>
  <App>...<CommandButton ... /></App>
</ThemeProvider>
```

This would let us inject the flag at a global level without passing it in manually.

**Comment by jakob101** [comment]
Can we help out with this? Where do we sign up?

**Comment by dzearing** [comment]
@cschleiden  and @micahgodbolt 

@jakob101 would you like to prototype `ThemeProvider`? I've started this and now that I'm done with the webpack dev server work, I'd like to spend more time with this and the css story. I think we'd have a ThemeProvider component, as well as a withTheme decorator/HOC. The HOC would pull any default overrides from context and both mix overrides into the passthrough props, but also re-export the merged context so nested ThemeProviders can work.

Then once that's there, I'd imagine this would work with any components without changes to the components.

The only thing I'm unsure of is how to get type safety in typescript. E.g.

```ts
componentDefaults: {
  CommandButton: {
    icon: 'Close'
  } as IButtonProps
}
```

For the specific prop `withoutGlobalClasses` after more thought, I almost want it to be an opt in thing, like "withLegacyClassNames" and have the default behavior to not expose them.

**Comment by cschleiden** [comment]
Sounds reasonable to me.

```TypeScript
componentDefaults: {
  CommandButton: {
    icon: 'Close'
  } as IButtonProps
}
```

Type safety for this will be tough, only option is probably to do something like this:

```TypeScript
interface componentDefaults: {
  CommandButton?: IButtonProps;
  Pivot?: IPivotProps;
}

`componentDefaults` would need to know about every component there is (potentially in different packages). 
```

**Comment by jakob101** [comment]
@dzearing I'll try to draft a proposal :) Let's talk about properties we need there.

**Comment by dzearing** [comment]
@jakob101 I have a prototype actually done. I should be able to send a PR tomorrow to see what you think.

**Comment by vandop** [comment]
Thanks for prioritizing this 👍

**Comment by dzearing** [comment]
I have this coded and working but can't submit a PR for this until tomorrow, need to add a few tests.

Also thinking about a better name than ThemeProvider, as this is really a way to inject customizations, which may extend beyond the definition of "theming". We may want to separate a ThemeProvider component which can generate a theme object which gets injected via this thing.

```tsx
<Customizer settings={ ... }>
  <Button />
<Customizer>
```

**Comment by jakob101** [comment]
If you're customizing more than colors, I agree a ```<Customizer>``` would make more sense. Otherwise it is just a theme, no?

**Comment by dzearing** [comment]
@Markionium submitted a change to introduce this recently:

```tsx
<Customizer settings={{ theme: { disableGlobalClassNames: true } }}>
  <Link ... />
</Customizer>
```

**Comment by micahgodbolt** [comment]
Is this issue still relevant? Is there anything left to be done?

**Comment by micahgodbolt** [comment]
This appears to be done. If it is not, tell me and I'll reopen.

**Comment by Markionium** [comment]
@micahgodbolt This is done. We are using this feature and it works fine :)
